// @kumar935/formfield/build/elements/FormField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

require('../styles/common.css');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactSelect = require('react-select');

var _reactSelect2 = _interopRequireDefault(_reactSelect);

var _Constants = require('./Constants');

var _StyledElements = require('./StyledElements');

var _ValidationProfiles = require('./ValidationProfiles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var selectStyleWrapperMap = {
  "default": _StyledElements.DefaultSelectWrapper,
  "lined": _StyledElements.LinedSelectWrapper
};

var FormField = function (_Component) {
  _inherits(FormField, _Component);

  function FormField() {
    _classCallCheck(this, FormField);

    var _this = _possibleConstructorReturn(this, (FormField.__proto__ || Object.getPrototypeOf(FormField)).call(this));

    _this.showHidePwd = function () {
      _this.setState(function (prevState) {
        if (_this.state.webKitTextSecurityPresent) {
          return {
            hide: !prevState.hide
          };
        } else {
          var hide = !prevState.hide;
          return {
            hide: hide,
            pwdType: hide ? 'password' : 'text'
          };
        }
      });
    };

    _this.handleSelectChange = function (newVal) {
      _this.setState({
        value: newVal,
        error: false,
        errorMsg: ''
      }, function () {
        if (_this.props.onChange) {
          _this.props.onChange(newVal);
        }
      });
    };

    _this.handleRadioChange = function (changeEvent) {
      _this.setState({
        value: changeEvent.target.value,
        error: false,
        errorMsg: ''
      });
    };

    _this.handleInputChange = function (e) {
      var newValue = e.target.value;
      if (_this.props.config.type === 'number') {
        if (isNaN(newValue)) return;
      }
      if (_this.props.config.sizeRange) {
        var max = _this.props.config.sizeRange[1];
        if (newValue.length > max) {
          return;
        }
      }
      if (_this.props.config.inputValidations) {
        var invalid = _this.isInputInvalid(newValue);
        if (invalid === true) return;
      }
      _this.setState({
        error: false,
        errorMsg: '',
        value: newValue
      }, function () {
        if (_this.props.config.liveValidations) _this.liveValidate();
        if (_this.props.onChange) {
          _this.props.onChange(newValue);
        }
      });
    };

    _this.onFocusCustomPwd = function () {
      _this.setState({ readOnly: false, focused: true });
    };

    _this.state = {
      error: false,
      errorMsg: '',
      value: '',
      pwdType: 'text',
      hide: false,
      readOnly: true,
      webKitTextSecurityPresent: false
    };
    return _this;
  }

  _createClass(FormField, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      console.log("hii? ", FormField.defaults);
      if (this.props.config.initVal) this.setState({ value: this.props.config.initVal });
      if (this.props.onRef) {
        this.props.onRef(this);
      } else {
        console.warn('define onRef function for: ' + this.props.config.label);
      }
      this.init();
    }
  }, {
    key: 'init',
    value: function init() {
      var _this2 = this;

      var isIE = /*@cc_on!@*/false || !!document.documentMode;
      if (this.props.config.type === 'custom-pwd') {
        var x = document.getElementsByTagName("input")[0];
        var style = window.getComputedStyle(x);
        if (style.webkitTextSecurity) {
          this.setState({ webKitTextSecurityPresent: true });
          window.setTimeout(function () {
            return _this2.setState({ hide: true });
          }, 100);
        } else {
          this.setState(_extends({
            pwdType: 'password',
            hide: true
          }, _extends({}, isIE ? { readOnly: false } : {})));
        }
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.setValue('');
      if (this.props.onRef) this.props.onRef(null);
    }
  }, {
    key: 'setError',
    value: function setError(config) {
      if (this.props.config.type === 'custom-pwd') this.setValue(this.value);
      this.setState({
        error: true,
        errorMsg: config.errorMsg || _Constants.ERROR_MSGS.DEFAULT
      });
    }
  }, {
    key: 'clearError',
    value: function clearError() {
      this.setState({
        error: false,
        errorMsg: ''
      });
    }
  }, {
    key: 'focus',
    value: function focus() {
      this.input.focus();
    }
  }, {
    key: 'isInputInvalid',
    value: function isInputInvalid(val) {
      var inputValidations = this.props.config.inputValidations || [];
      var invalid = false;
      for (var i = 0; i < inputValidations.length; i++) {
        var valProfile = inputValidations[i];
        if (FormField.defaults.validationProfiles[valProfile] === undefined) {
          if (valProfile.indexOf(_Constants.REGEXP_PREFIX) === 0) {
            var regExp = new RegExp(valProfile.replace(_Constants.REGEXP_PREFIX, ""));
            invalid = !regExp.test(this.state.value);
            if (invalid) break;
          }
        } else {
          var result = FormField.defaults.validationProfiles[inputValidations[i]](val, this.props);
          if (!result.valid) {
            invalid = true;
            break;
          }
          invalid = false;
        }
      }
      return invalid;
    }
  }, {
    key: 'liveValidate',
    value: function liveValidate() {
      var validationProfiles = this.props.config.liveValidations || ['default'];
      return this.validate(validationProfiles);
    }
  }, {
    key: 'isValid',
    value: function isValid() {
      var validationProfiles = this.props.config.validationProfiles || ['default'];
      return this.validate(validationProfiles);
    }
  }, {
    key: 'validate',
    value: function validate(validationProfiles) {
      var valid = false;
      if (validationProfiles.length === 0) valid = true;
      for (var i = 0; i < validationProfiles.length; i++) {
        var valProfile = validationProfiles[i];
        if (FormField.defaults.validationProfiles[valProfile] === undefined) {
          if (valProfile.indexOf(_Constants.REGEXP_PREFIX) === 0) {
            var regExp = new RegExp(valProfile.replace(_Constants.REGEXP_PREFIX, ""));
            valid = regExp.test(this.state.value);
            if (!valid) {
              this.setState({
                error: true,
                errorMsg: _Constants.ERROR_MSGS.INVALID_INPUT
              });
              break;
            }
          }
        } else {
          var result = FormField.defaults.validationProfiles[valProfile](this.state.value, this.props);
          if (!result.valid) {
            this.setState({
              error: true,
              errorMsg: result.errorMsg
            });
            valid = false;
            break;
          }
          valid = true;
        }
      }
      return valid;
    }
  }, {
    key: 'getValue',
    value: function getValue() {
      if (this.props.config.type === 'custom-pwd') return this.getValueSetBlank();
      return this.state.value;
    }
  }, {
    key: 'getOnlyValue',
    value: function getOnlyValue() {
      return this.state.value;
    }
  }, {
    key: 'setValue',
    value: function setValue(value) {
      var _this3 = this;

      return new Promise(function (resolve) {
        _this3.setState(_extends({
          value: value
        }, !value ? {} : { error: false, errorMsg: "" }), function () {
          resolve({ "done": "ok" });
        });
      });
    }
  }, {
    key: 'getValueSetBlank',
    value: function getValueSetBlank() {
      this.value = this.state.value;
      this.setValue('');
      return this.value;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      var _props = this.props,
          config = _props.config,
          onRef = _props.onRef,
          rest = _objectWithoutProperties(_props, ['config', 'onRef']);

      var common = {
        color: FormField.defaults.color,
        ref: function ref(input) {
          _this4.input = input;
        }
      };
      if (config.type === 'radio') {
        return _react2.default.createElement(
          'div',
          { className: 'radio-group' },
          config.label && !config.hideLabel ? _react2.default.createElement(
            _StyledElements.Label,
            null,
            config.label
          ) : '',
          _react2.default.createElement(
            'div',
            { className: 'radio-group-wraper' },
            config.options.map(function (option) {
              return _react2.default.createElement(
                _react2.default.Fragment,
                null,
                _react2.default.createElement(
                  _StyledElements.Label,
                  { className: 'radio' },
                  _react2.default.createElement(_StyledElements.Input, {
                    type: 'radio',
                    color: FormField.defaults.color,
                    value: option.value,
                    checked: _this4.state.value === option.value,
                    onChange: _this4.handleRadioChange }),
                  _react2.default.createElement('span', null)
                ),
                _react2.default.createElement(
                  _StyledElements.Label,
                  { className: 'radioLabel', htmlFor: option.label },
                  option.label
                )
              );
            })
          ),
          _react2.default.createElement(
            'span',
            { className: 'error' },
            this.state.errorMsg
          )
        );
      }
      if (config.type === 'select') {
        if (FormField.defaults.styles.selectStyle === "default" || !FormField.defaults.styles.selectStyle) {
          require('react-select/dist/react-select.css');
        }
        var SelectWrapper = selectStyleWrapperMap[FormField.defaults.styles.selectStyle] || _StyledElements.DefaultSelectWrapper;
        var selectProps = _extends({
          name: 'select-' + config.label,
          value: this.state.value,
          options: config.options || [],
          onChange: this.handleSelectChange,
          clearable: false,
          noResultsText: config.loading ? 'Loading...' : 'No Results'
        }, config.SelectArgs ? config.SelectArgs : {});
        return _react2.default.createElement(
          SelectWrapper,
          {
            className: 'select-box label ' + (config.containerClass || ''),
            error: this.state.error.toString(),
            color: FormField.defaults.color },
          _react2.default.createElement(
            _StyledElements.Label,
            null,
            config.label
          ),
          this.state.disabled ? _react2.default.createElement(_reactSelect2.default, _extends({}, selectProps, {
            disabled: true
          })) : _react2.default.createElement(_reactSelect2.default, _extends({}, selectProps, {
            disabled: false
          })),
          _react2.default.createElement(
            'span',
            { className: 'error-msg select-error' },
            this.state.errorMsg
          ),
          FormField.defaults.styles.selectStyle === "lined" ? _react2.default.createElement(_StyledElements.Line, null) : ''
        );
      }
      return _react2.default.createElement(
        _StyledElements.InputWrapper,
        { className: config.containerClass || '' },
        function () {
          switch (config.type) {
            case 'text':
            case 'password':
              return _react2.default.createElement(_StyledElements.Input, _extends({
                type: config.type,
                value: _this4.state.value,
                error: _this4.state.error.toString(),
                autoComplete: 'off'
              }, rest, {
                onChange: _this4.handleInputChange,
                required: true
              }, common));
              break;
            case 'custom-pwd':
              return [_react2.default.createElement(_StyledElements.Input, { type: 'password', tabIndex: '-1', required: true, name: 'fake_pwd', style: { position: 'absolute', top: '-20000px' }, key: 'fake-pwd' }), _react2.default.createElement(_StyledElements.Input, _extends({
                key: 'real-pwd',
                type: _this4.state.pwdType
              }, _this4.state.hide ? { 'data-type': 'custom-pwd' } : {}, {
                value: _this4.state.value
              }, _this4.state.readOnly ? { 'readOnly': 'true' } : {}, {
                onFocus: _this4.onFocusCustomPwd,
                autoComplete: 'off',
                error: _this4.state.error.toString()
              }, rest, {
                onChange: _this4.handleInputChange,
                required: true
              }, common))];
              break;
            case 'number':
              return _react2.default.createElement(_StyledElements.Input, _extends({
                type: 'text',
                value: _this4.state.value,
                error: _this4.state.error.toString()
              }, rest, {
                onChange: _this4.handleInputChange,
                required: true
              }, common));
              break;
            default:
              return _react2.default.createElement(_StyledElements.Input, _extends({
                type: 'text',
                value: _this4.state.value,
                autoComplete: 'off',
                error: _this4.state.error.toString()
              }, rest, {
                onChange: _this4.handleInputChange,
                required: true
              }, common));
          }
        }(),
        config.type === 'custom-pwd' ? _react2.default.createElement('span', { className: 'eye ' + (this.state.hide ? 'show' : 'hide'), onClick: this.showHidePwd }) : '',
        _react2.default.createElement(
          _StyledElements.Label,
          { htmlFor: this.props.id },
          this.props.config.label
        ),
        _react2.default.createElement(
          'span',
          { className: 'error-msg' },
          this.state.errorMsg
        ),
        config.posRightBtn && config.posRightBtn.domClass ? _react2.default.createElement(_StyledElements.Label, {
          className: 'position-right ' + config.posRightBtn.domClass,
          onClick: config.posRightBtn.onClick
        }) : '',
        _react2.default.createElement(_StyledElements.Line, { className: 'line', color: FormField.defaults.color })
      );
    }
  }]);

  return FormField;
}(_react.Component);

FormField.defaults = {
  styles: _StyledElements.styleDefaults,
  validationProfiles: _ValidationProfiles.validationProfilesConfig
};
exports.default = FormField;
